src/tickets/utils.ts

```diff
@@ -14,30 +26,112 @@ export function isPreactivatedTicket(
   );
 }
 
-export const filterActiveFareContracts = (fareContracts: FareContract[]) => {
-  const isValidNow = (f: FareContract): boolean => {
-    const firstTravelRight = f.travelRights?.[0];
-    if (isPreactivatedTicket(firstTravelRight)) {
-      return firstTravelRight.endDateTime.toMillis() > Date.now();
-    }
+export function isSingleTicket(
+  travelRight: TravelRight | undefined,
+): travelRight is PreactivatedSingleTicket {
+  return travelRight?.type === 'PreActivatedSingleTicket';
+}
+
+function isOrWillBeActivatedFareContract(f: FareContract): boolean {
+  return (
+    f.state === FareContractState.Activated ||
+    f.state === FareContractState.NotActivated
+  );
+}
+
+function isValidPreactivatedTicket(ticket: PreactivatedTicket): boolean {
+  return ticket.endDateTime.toMillis() > Date.now();
+}
+
+function hasValidCarnetTicket(tickets: CarnetTicket[]): boolean {
+  const {usedAccesses} = flattenCarnetTicketAccesses(tickets);
+  const [lastUsedAccess] = usedAccesses?.slice(-1);
+  const validTo = lastUsedAccess?.endDateTime.toMillis() ?? 0;
+  return Date.now() < validTo;
+}
+
+export function isValidRightNowFareContract(f: FareContract): boolean {
+  if (!isOrWillBeActivatedFareContract(f)) {
     return false;
+  }
+
+  const firstTravelRight = f.travelRights?.[0];
+  if (isPreactivatedTicket(firstTravelRight)) {
+    return isValidPreactivatedTicket(firstTravelRight);
+  } else if (isCarnetTicket(firstTravelRight)) {
+    return hasValidCarnetTicket(f.travelRights.filter(isCarnetTicket));
+  }
+
+  return false;
+}
+
+function hasActiveOrUsableCarnetTicket(tickets: CarnetTicket[]): boolean {
+  const [firstTicket] = tickets;
+  const {
+    usedAccesses,
+    maximumNumberOfAccesses,
+    numberOfUsedAccesses,
+  } = flattenCarnetTicketAccesses(tickets);
+
+  const [lastUsedAccess] = usedAccesses?.slice(-1);
+  const validTo = lastUsedAccess?.endDateTime.toMillis() ?? 0;
+  const now = Date.now();
+  return (
+    firstTicket.endDateTime.toMillis() > now &&
+    (maximumNumberOfAccesses > numberOfUsedAccesses || now < validTo)
+  );
+}
+
+type FlattenedCarnetTicket = {
+  usedAccesses: CarnetTicketUsedAccess[];
+  maximumNumberOfAccesses: number;
+  numberOfUsedAccesses: number;
+};
+
+export function flattenCarnetTicketAccesses(
+  tickets: CarnetTicket[],
+): FlattenedCarnetTicket {
+  const allUsedAccesses = tickets.map((t) => t.usedAccesses ?? []);
+  const usedAccesses = flatten(allUsedAccesses).sort(
```

Comment:
For egen læring: Er det noe i JavascriptCore som gjør at man ikke får brukt ES2019-funksjonalitet som `flatMap`?